Ignore bad NetDefs and files via parser flags

daniloegea opened the pull request:
## Description

This is an implementation of what I called "parser flags". The idea is to enable the user to change some parsing decisions.

The main application at the moment is to support ignoring parsing errors. Currently Netplan will not generate any configuration if a little mistake is made in one of its YAML files. It can be really bad if you're operating a remote system and end up rebooting the system with a syntax issue in one of your files. Your system will basically boot without any network configuration.

The IGNORE_ERRORS flag will allow the Netplan generator to ignore bad files and bad netdefs so it will still generate some configuration.

**TESTS**

Integration tests will be executed twice now, once without the ignore_errors flags and once with it enabled. It's intended to increase the confidence that the parser behaves as expected when the flag is enabled and the configuration is good.

I prepared a PPA for Ubuntu Noble with this patch: https://launchpad.net/~danilogondolfo/+archive/ubuntu/netplan-parser-flags

Suggestions of tests:

It can be easily tested in a LXD VM. You can add some broken configuration and reboot the VM or call `/usr/libexec/netplan/generate -i` to see the parsing process when it's ignoring errors

1. Break the default `50-cloud-init.yaml` and reboot the VM. For example, the config below should still bring enp5s0 up and with DHCP working.

```yaml
network:
  version: 2
  ethernets:
    enp5s0:
      dhcp4: true
      dhcp: true
```

2. Add a second file that is not a valid YAML
3.  Add a third broken file with some complex configuration and dependencies between netdefs. For example, with the configuration below `eth2` should still have a backend configuration file and it should contain `Bond=bond0`.
```yaml
network:
  ethernets:
    eth0: {}
    eth1: {}
    eth2:
      a: b

  bonds:
    bond0:
      interfaces:
        - eth0
        - eth1
        - eth2
```
4. Add a fourth broken file with missing dependencies. For example, with the config below you still should have a `br0` interface created:
```yaml
network:
  bridges:
    br0:
      addresses:
        - 10.20.30.40/24

      interfaces:
        - aaaa
        - bbbb
        - dddd
```

## Checklist

- [ ] Runs `make check` successfully.
- [ ] Retains 100% code coverage (`make check-coverage`).
- [ ] New/changed keys in YAML format are documented.
- [ ] \(Optional\) Adds example YAML for new feature.
- [ ] \(Optional\) Closes an open bug in Launchpad.

schopin-pro commented on python-cffi/netplan/_build_cffi.py:
```diff
@@ -68,6 +68,7 @@
     // Parser
     NetplanParser* netplan_parser_new();
     void netplan_parser_clear(NetplanParser **npp);
+    void netplan_parser_set_flags(NetplanParser *npp, int flags);
```
suggestion: we probably want to expose a `netplan_parser_get_flags(npp)` already.

This would give us the liberty to modify the flags as side effects of other operations, like when we're pre-loading data, etc... as the caller wouldn't need to keep an external copy of the flags and could just do `netplan_parser_set_flags(npp, netplan_parser_get_flags(npp) | NETPLAN_PARSER_IGNORE_ERRORS);`

schopin-pro commented on include/parse.h:
```diff
@@ -53,6 +53,15 @@ netplan_parser_reset(NetplanParser *npp);
+ * @param[in] flags The value of the flags. The possible values are defined in @ref NETPLAN_PARSER_FLAGS
+ */
+NETPLAN_PUBLIC void
+netplan_parser_set_flags(NetplanParser* npp, int flags);
```
suggestion: fail on unknown flags

(making the suggestion at the header level as this is API design)

I think it'd make sense to fail loudly when seeing an unknown flag rather than silently ignore it.

slyon replied:
+1 this should return `gboolean` and take a ` NetplanError** error` argument to be filled with details in case of failure.

schopin-pro commented on python-cffi/netplan/parser.py:
```diff
@@ -13,12 +13,17 @@
 
 
+class NETPLAN_PARSER_FLAGS(IntEnum):
+    IGNORE_ERRORS = 1
```
suggestion: Just call this class `Flags`

We're already in the `netplan.parser` namespace!

schopin-pro commented on python-cffi/netplan/parser.py:
```diff
@@ -46,3 +51,6 @@ def load_nullable_fields(self, input_file: IO):
                                  self._ptr, input_file.fileno(), constraint.encode('utf-8'))
+
+    def set_flags(self, flags: int):
+        lib.netplan_parser_set_flags(self._ptr, flags)
```
suggestion: use a property instead of an actual call

Assuming previous suggestions are implemented, that'd make the following possible:

```py
npp.flags |= Flags.IGNORE_ERRORS
```
rather than
```py
npp.set_flags(saved_flags | NETPLAN_PARSER_FLAGS.IGNORE_ERRORS)
```

schopin-pro commented on src/parse.c:
```diff
@@ -3134,6 +3134,23 @@ node_is_nulled_out(yaml_document_t* doc, yaml_node_t* node, const char* key_pref
+    }
+
+    reset_netdef(netdef, NETPLAN_DEF_TYPE_NONE, NETPLAN_BACKEND_NONE);
+    g_free(netdef);
```
question: wouldn't that create dangling pointers?

IIRC we have cross-references on netdefs in cases of bridges and the likes, and we try to resolve them on the spot. If that is still true, we should at least add big fat warnings to the docs of this function so that our future selves don't start using it wrongly. Conversely, since we're only dropping the currently-parsed netdef for now we could actually make this function about that one special case?

Related question: don't we write backreferences *in* previously defined netdefs in some cases? It's been a long while since I've looked at the parser code, but I dimly remember that in some cases we go back and modify existing netdefs while parsing other ones, possibly to add a pointer to the new one.

daniloegea replied:
Yes, definitely. In the end I didn't do it. I decided to leave the netdef "half parsed" in the parser state in case of failure. It makes things simpler and increase the chances you'll have a working configuration in the end (I think).

If we end up generating invalid back end configuration due to this, both NM and networkd will reject the file so it should be fine.

schopin-pro reviewed (commented):
Well, that was a fun lunch break :)

slyon commented on src/generate.c:
```diff
@@ -44,10 +44,12 @@ static gchar** files;
 static GOptionEntry options[] = {
     {"root-dir", 'r', 0, G_OPTION_ARG_FILENAME, &rootdir, "Search for and generate configuration files in this root directory instead of /", NULL},
     {G_OPTION_REMAINING, 0, 0, G_OPTION_ARG_FILENAME_ARRAY, &files, "Read configuration from this/these file(s) instead of /etc/netplan/*.yaml", "[config file ..]"},
+    {"ignore-errors", 'i', 0, G_OPTION_ARG_NONE, &ignore_errors, "Ignores files and/or interfaces that fail parsing.", NULL},
```
nitpick: We shouldn't call it "interfaces". Maybe use "network definitions" or "network definition blocks" instead?

slyon commented on src/parse.c:
```diff
@@ -3245,8 +3258,29 @@ handle_network_type(NetplanParser* npp, yaml_node_t* node, const char* key_prefi
+                gchar* warning_msg = NULL;
+                if (error && *error)
+                    warning_msg = (*error)->message;
+                g_warning("Skipping interface due to parsing errors. %s: %s", scalar(key), warning_msg);
```
nitpick: "Skipping network definition ..."

slyon commented on include/types.h:
```diff
@@ -164,3 +164,10 @@ enum NETPLAN_EMITTER_ERRORS {
+ * @brief   Flags used to change the parser behavior.
+ */
+enum NETPLAN_PARSER_FLAGS {
+    NETPLAN_PARSER_IGNORE_ERRORS = 1, ///< Ignore parsing errors such as bad YAML files and definitions.
```
nitpick: Maybe we should define this as `0x1` (or `0b1` even) to make it more obvious that this is a flag/bit.

schopin-pro replied:
I've seen `1 << 0` (and then `1 << 1`, etc), which means you don't even need to know your powers of 2 nor count the number of zeros ;).

slyon commented on src/util.c:
```diff
@@ -656,11 +656,21 @@ netplan_parser_load_yaml_hierarchy(NetplanParser* npp, const char* rootdir, GErr
+                if (error && *error)
+                    warning_msg = (*error)->message;
+
+                g_warning("Skipping YAML file due to parsing errors. %s", warning_msg);
```
todo: We should mention the specific filename + path here, too.

daniloegea replied:
I trust it will be present in the error->message in case of failure...

slyon commented on tests/generator/test_errors.py:
```diff
@@ -999,3 +999,76 @@ def test_nm_only_supports_unicast_routes(self):
+  ethernets:
+    eth123:
+      dhcp4: true
+    eth321:
+      dhcp4: 1''', confs={'b': '''network:
+  bridges:
+    br0:
+      interfaces: [ eth123 ]
+    br1:
+      interfaces: [ eth321 ]'''}, expect_fail=False, skip_generated_yaml_validation=True, ignore_errors=True)
```
question: This is the case Simon described above IIUC. So we would still generate configuration for `br1`, even though `eth321` is broken. That would cause `br1` being unable to be configured at runtime, whith correspondig errors in journal, right?

daniloegea replied:
Yes, in the end I didn't drop netdefs from the state, so it will generate configuration for all the interfaces. If they are broken because they are half parsed or not validated, I trust the back end will ignore the config file we generate.

slyon reviewed (commented):
I like the overall approach of this. It is a slight change in behavior, but only for the better, as a non-configured network (due to parsing errors) is always a bad thing. It could lead to unexpected network configuration if some files are ignored but other aren't, but IMO that is still better than no network configuration at all.

daniloegea commented:
Thanks for your comments, lads. I basically adopted all of them.

I decided to not drop bad netdefs anymore. I think it's actually better to keep them and generate configuration for what was parsed until the point it failed. There is a chance the interface will still be brought up and work if the error happened after settings IPs or DHCP for example.

To make sure (or at least increase the confidence) that these changes will not cause a change in behavior when the flag is enabled and all the configuration is good, I changed our `integration/run.py` script to run each test twice, with and without the flag.

The new "Configuration fuzzing" CI action is catching a memory leak! As far as I can tell, it's happening inside glib and there is  a comment in the code that handles datalists saying it could happen, see FR-5666 for more details. So, this leak is not introduces by these changes.

I prepared a [PPA](https://launchpad.net/~danilogondolfo/+archive/ubuntu/netplan-parser-flags) and added to suggestions of tests in the PR comments.

slyon commented on src/types-internal.h:
```diff
@@ -263,6 +263,9 @@ struct netplan_parser {
     GHashTable* null_overrides;
     GHashTable* global_renderer;
+
+    /* Flags used to change the parser's behavior */
+    unsigned int flags;
```
suggestion (non-blocking): I'd love to see an additional boolean field that gets set whenever any parsing actually failed. It should get carrier over from NetplanParser into NetplanState, so we can check the final state, e.g. `State.parsing_errors` (not sure about the naming, yet).

daniloegea replied:
I like the idea. I just used a different approach, I added a counter to keep track of how many errors were ignored. I added it in two separate commits to make it easier to review.

slyon commented on tests/generator/test_errors.py:
```diff
@@ -999,3 +999,268 @@ def test_nm_only_supports_unicast_routes(self):
+      dhcp4: false''', expect_fail=False, skip_generated_yaml_validation=True, ignore_errors=True)
+
+        self.assertTrue(self.file_exists('10-netplan-eth0.network'))
+        self.assertTrue(self.file_exists('10-netplan-eth1.network'))
```
nitpick (non-blocking): Should we check that the `eth1.network` config does not contain any DHCP config here?

daniloegea replied:
I think it's not necessary. There is a test in `test_netplan_parser.c:test_parser_flags_ignore_errors` that checks if `dhcp4` is still false after a failure. It should be enough to check if the netdef field has the expected value I guess, instead of checking if the configuration for it was generated or not...

slyon commented on tests/generator/test_errors.py:
```diff
@@ -999,3 +999,268 @@ def test_nm_only_supports_unicast_routes(self):
+    eth0: {}
+  bonds:
+    bond0:
+      interfaces: [eth0]
+      badkey: badvalue''', expect_fail=False, skip_generated_yaml_validation=True, ignore_errors=True)
```
question: I assume this would still be fine if we define `badkey: badvalue` first, before `interfaces: [...]`?

daniloegea replied:
If the error happened before `interfaces: [eth0]` this fields wouldn't be parsed. The `bond0` would still be there, but with an empty list of members.

slyon replied:
Thanks for the explanation, this should be fine!

slyon commented on src/generate.c:
```diff
@@ -195,11 +197,13 @@ find_interface(gchar* interface, GHashTable* netdefs)
         error_code = 1; \
         fprintf(stderr, "%s\n", error->message); \
         goto cleanup;\
+    } else if (error && ignore_errors) {\
+        g_clear_error(&error); \
```
suggestion: We should consider still logging the error here, as least to debug log.. Maybe using an `IGNORED: ...` prefix?

slyon commented on src/networkd.c:
```diff
@@ -620,12 +620,14 @@ write_netdev_file(const NetplanNetDefinition* def, const char* rootdir, const ch
+                    g_string_free(s, TRUE);
+                    return;
+                }
+                g_string_append_printf(s, "Kind=veth\n\n[Peer]\nName=%s\n", def->veth_peer_link->id);
```
nitpick: I wonder if we should still write the `Kind=veth\n` line, even if we don't have `[Peer].Name` defined?

slyon reviewed (commented):
Thanks a lot for working through all the remarks. This should be relatively safe now.

The doubling of testing-time is a though sell... OTOH more coverage is better. But at least it should be easy to disable (e.g. through ENV variable), so we can skip the double-testing for local tests.

I left a bunch of additional inline comments that need additional consideration.

daniloegea commented on src/parse.c:
```diff
@@ -3729,6 +3754,21 @@ netplan_parser_clear(NetplanParser** npp_p)
 }
 
+gboolean
+netplan_parser_set_flags(NetplanParser* npp, const unsigned int flags)
```
Right, I added it. I wasn't convinced it was necessary as it's a simple operation. But in the future we might have more flags and want to report more details about the error.

daniloegea commented on src/generate.c:
```diff
@@ -243,6 +245,9 @@ int main(int argc, char** argv)
     }
 
     npp = netplan_parser_new();
+    if (ignore_errors || called_as_generator)
```
As `netplan generate` will not include the `--ignore-errors` (only the generator), I just added a short comment about it there. I added a new explanation section about the generator.

daniloegea commented:
Thanks Lukas, I tried to address all of your comments.

I added an error counter to the parser state instead of a boolean, that might be more useful I think.
I also added a new parser_error to the enum and exposed it to the python code.
I will squash the two extra commits that contain these changes later.

The first commit was split into the definition and implementation of flags as suggested by @schopin-pro in a conversation we had.

The config_fuzzer was changed to run the generator with the `--ignore-errors` flag for one YAML at a time to avoid that annoying glib memory leak. It's not the ideal test though... I need to investigate how the leak can be identified and ignored.

slyon reviewed (approved):
Thanks for addressing the remarks!

I like the error handling to be integrated with existing error handling logic and the `.error_count` field.
